pkg/metrics/metrics.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2018 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package metrics contains global structures related to metrics collection
+// cert-manager exposes the following metrics:
+// certificate_expiration_seconds{name, namespace}
+package metrics
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	corelisters "k8s.io/client-go/listers/core/v1"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/util/errors"
+	"github.com/jetstack/cert-manager/pkg/util/kube"
+)
+
+const (
+	// Namespace is the namespace for cert-manager metric names
+	namespace                              = "certmanager"
+	prometheusMetricsServerAddress         = "0.0.0.0:9402"
+	prometheusMetricsServerShutdownTimeout = 5 * time.Second
+	prometheusMetricsServerReadTimeout     = 8 * time.Second
+	prometheusMetricsServerWriteTimeout    = 8 * time.Second
+	prometheusMetricsServerMaxHeaderBytes  = 1 << 20 // 1 MiB
+)
+
+// Default set of metrics
+var Default = New()
+
+var CertificateExpiryTimeSeconds = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "certificate_expiration_seconds",
```

Comment:
Can we rename to `certificate_expiration_timestamp_seconds` to make it as clear as possible that this is a unix timestamp value?

From reading the Prometheus docs on this stuff, I think this metric *may* be better suited as a 'summary' metric. But let's role with this for now, with the adjusted name 😄